### DISCUSSION TITLE
No mechanism to customize the name of the kubernetes serviceaccount generated by serviceaccounttemplate

### DISCUSSION THREAD

**Discussion by apjoseph** [cloudnative-pg/cloudnative-pg#4219] [Ideas]
There appears to be no way to customize the name of the database cluster service account generated with `serviceaccounttemplate`. Current behavior appears to create a service account with the same name as the cluster. The [documentation for AWS EKS IRSA backups](https://cloudnative-pg.io/documentation/1.20/backup_recovery/#iam-role-for-service-account-irsa) shows an example of adding the `eks.amazonaws.com/role-arn` annotation, but IRSA requires the trust policy for the IAM role to contain the name of the generated serviceaccount:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Federated": "arn:aws:iam::111122223333:oidc-provider/oidc.eks.region-code.amazonaws.com/id/EXAMPLED539D4633E53DE1B71EXAMPLE"
            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "StringEquals": {
                    "oidc.eks.region-code.amazonaws.com/id/EXAMPLED539D4633E53DE1B71EXAMPLE:sub": "system:serviceaccount:default:my-service-account",
                    "oidc.eks.region-code.amazonaws.com/id/EXAMPLED539D4633E53DE1B71EXAMPLE:aud": "sts.amazonaws.com"
                }
            }
        }
    ]
}
```

It would be great to be able to customize the serviceaccountname to prevent conflicts -but at the very least, the current behavior should be documented so that users of the operator know what name to put in the IAM role's trust policy.

**Comment by dfsdevops** [comment]
I would like it even more if I could specify a pre-existing serviceaccount to use. I use terraform to bootstrap the serviceaccounts to utilize IRSA. Having to set this annotation in the cluster object is less convenient because I don't use terraform for that.

**Comment by sxd** [comment]
Turning this into a discussion, since it's an open matter not a clear issue

**Comment by jlmartins-sf** [comment]
It would be a great feature...
In a case of restore backup, it could be possible to restore using a specific name of serviceAccount that uses ARN from AWS, using a universal serviceAccount only to test backups weekly, for example:

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Federated": "arn:aws:iam::00000000000:oidc-provider/oidc.eks.us-east-1.amazonaws.com/id/423D3224234ED232D32D23D"
            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "ForAnyValue:StringEquals": {
                    "oidc.eks.us-east-1.amazonaws.com/id/423D3224234ED232D32D23D:sub": [
                        "system:serviceaccount:databases:cnpg-sa-backup-restore"
                    ],
                    "oidc.eks.us-east-1.amazonaws.com/id/423D3224234ED232D32D23D:aud": "sts.amazonaws.com"
                }
            }
        }
    ]
}
```

**Comment by mtb-xt** [comment]
This is something I'd expect CNPG to have, for example, if we want to use one serviceaccount to run backups on several cnpg clusters. Right now, I'd have to set my IAM role's trust policy to trust all service accounts

**Comment by hervenicol** [comment]
I'd love to be able to use an existing serviceaccount as well!

**Comment by calizarr** [comment]
I've added an issue with a proposed PR fix over here for both arbitrary ServiceAccount name and the ability to use an existing ServiceAccount.
https://github.com/cloudnative-pg/cloudnative-pg/issues/8909

https://github.com/cloudnative-pg/cloudnative-pg/pull/8921
